doc/specs/#492 - Default Terminal/spec.md

```diff
@@ -0,0 +1,248 @@
+
+The goal is to offer the ultimate in choice here where any of the components can be replaced as necessary for a 1st or 3rd party scenario.
+
+### Overview
+
+#### Inbox console
+
+The inbox console will be updated to support delegation of the incoming console client application connection to another console API server if one is registered and available.
+
+We leave the inbox console in-place and always available on the operating system for these reasons:
+
+1. A last chance fall-back should any of the delegation operations fail
+1. An ongoing host for applications that aren't going to need a window at all
+1. Continued support of our legacy `conhostv1.dll` environment, if chosen
+
+The general operation is as follows:
+
+- A command-line client application is started (from the start menu, run box, or any other `CreateProcess` or `ShellExecute` route) without an existing console server attached
+- The inbox console is launched from `C:\windows\system32\conhost.exe` as always by the initialization routines inside `kernelbase.dll`.
+- The inbox console accepts the incoming initial connection and looks for the `ShowWindow` information on the connection packet, as received from the kernel's process creation routines based on the parameters given to the `CreateProcess` call. (See [CREATE_NO_WINDOW](https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags) flag for details.)
+- If the session is about to create a window, check for registration of a delegated/updated console and hand-off to it if it exists.
+- Otherwise, start normally.
+
+This workflow affords us several benefits:
+
+- The only inbox component we have to change is `conhost.exe`, the one we already regularly update from open source on a regular basis. There is no change to the `kernelbase.dll` console initialization routines, `conclnt.lib` communication protocol, nor the `condrv.sys` driver.
+- We should be able to make this change quickly, relatively easily, and the code delta should be relatively small
+    - This makes it easy to squeeze in early in the development of the solution and get it into the Windows OS product as soon as possible for self-hosting, validation, and potentially shipping in the OS before the remainder of the solution has shaken out
+    - This also makes it potentially possible to backport this portion of the code change to popular in-market versions of Windows 10. For instance, WSL2 has just backported to 1903 and 1909. The less churn and risk, the easier it is to sell backporting.
+
+**To be decided:**
+- If no updated console exists, potentially check for registration of a terminal UX that is willing to use the inbox ConPTY bits, start it, and transition to being a PTY instead.
+
+The registration would operate as follows:
+- A registry key in `HKCU\Console` (format `REG_SZ`, name `DelegationConsole`) would specify the path to the replacement console that would be used to service the remainder of the connection process.

```

Comment:
Perhaps the key should be in a sub-key that could be further ACL'd later to prevent stomping? That would potentially collide with the LNK mapping, but not super likely.

Alternatively, we could use the HKLM paths we have, but that feels bad because apps can be installed per user and your terminal is really a user choice thing.

Reply:
I'm mildly worried that we're allowing a process to set a reg key to a random binary that we'll call a function in _on every console launch_. Seems like a trivial way for an application to sniff _all_ console messages, by just installing `sniffer.exe` as the delegated console, that then later just calls openconsole.exe (to make it seem like nothing's wrong).

HKCU seems to make sense, but I do think we need some way of preventing apps from just arbitrarily writing this themselves

Reply:
In the prototype, based on this feedback, I moved it down a node and intend to secure it with an ACL such that only the Settings UI page we're gonna make can access it (or someone with admin rights, probably). Alternately, if Settings UI won't let me keep it there, I'll put it in whatever HKCU/HKLM node is valid per Settings UI specs.